sql/core/src/main/scala/org/apache/spark/sql/execution/python/EvalPythonUDTFExec.scala

```diff
@@ -137,4 +150,46 @@ trait EvalPythonUDTFExec extends UnaryExecNode {
       }
     }
   }
+
+  lazy val memoryConsumer: Option[PythonUDTFMemoryConsumer] = {
+    if (TaskContext.get() != null) {
+      Some(PythonUDTFMemoryConsumer(udtf))
+    } else {
+      None
+    }
+  }
+}
+
+/**
+ * This class takes responsibility to allocate execution memory for UDTF evaluation before it begins
+ * and free the memory after the evaluation is over.
+ *
+ * Background: If the UDTF's 'analyze' method returns an 'AnalyzeResult' with a non-empty
+ * 'acquireExecutionMemoryMb' value, this value represents the amount of memory in MB that
+ * the UDTF should request from each Spark executor that it runs on. Then the UDTF takes
+ * responsibility to use at most this much memory, including all allocated objects. The purpose of
+ * this functionality is to prevent executors from crashing by running out of memory due to the
+ * extra memory consumption invoked by the UDTF's 'eval' and 'terminate' and 'cleanup' methods.
+ *
+ * In this class, Spark calls 'TaskMemoryManager.acquireExecutionMemory' with the requested number
+ * of MB, and when Spark calls __init__  of the UDTF later, it updates the acquiredExecutionMemory
+ * integer passed into the UDTF constructor to the actual number returned from
+ * 'TaskMemoryManager.acquireExecutionMemory', so the 'eval' and 'terminate' and 'cleanup' methods
+ * know it and can ensure to bound memory usage to at most this number.
+ */
+case class PythonUDTFMemoryConsumer(udtf: PythonUDTF)
+  extends MemoryConsumer(TaskContext.get().taskMemoryManager(), MemoryMode.ON_HEAP) {
```

Comment:
I'm not sure about this and how this helps to avoid OOM.
IIUC, `MemoryConsumer` is supposed to manage the JVM memory, and configured as `ON_HEAP`. The on-heap memory should be reserved by JVM and won't be reduced once it's reserved.

Reply:
My mistake, I should have written `MemoryMode.OFF_HEAP` here to indicate that the memory consumed by the Python subprocess is outside the JVM. Fixed this.